rust/worker/src/execution/data/data_chunk.rs

```diff
@@ -0,0 +1,146 @@
+use std::ops::Range;
+use std::sync::Arc;
+
+use crate::types::EmbeddingRecord;
+
+#[derive(Clone, Debug)]
+pub(crate) struct DataChunk {
+    data: Arc<[EmbeddingRecord]>,
+    visibility: Arc<[bool]>,
+}
+
+impl DataChunk {
+    pub fn new(data: Arc<[EmbeddingRecord]>) -> Self {
+        let len = data.len();
+        DataChunk {
+            data,
+            visibility: vec![true; len].into(),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn visible_len(&self) -> usize {
+        // return all the data with visibility true
+        self.visibility.iter().filter(|&v| *v).count()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    pub fn get(&self, index: usize) -> Option<&EmbeddingRecord> {
+        if index < self.data.len() {
+            Some(&self.data[index])
+        } else {
+            None
+        }
+    }
+
+    pub fn get_visibility(&self, index: usize) -> Option<bool> {
+        if index < self.visibility.len() {
+            Some(self.visibility[index])
+        } else {
+            None
+        }
+    }
+
+    pub fn iter(&self) -> DataChunkIteraror<'_> {
+        DataChunkIteraror {
+            chunk: self,
+            index: 0,
+        }
+    }
+
+    pub fn set_visibility(&mut self, visibility: Vec<bool>) {
+        self.visibility = visibility.into();
+    }
+}
+
+pub(crate) struct DataChunkIteraror<'a> {
+    chunk: &'a DataChunk,
+    index: usize,
+}
+
+impl<'a> Iterator for DataChunkIteraror<'a> {
+    type Item = (&'a EmbeddingRecord, usize);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.index < self.chunk.len() {
+            let index = self.index;
+            if self.chunk.get_visibility(index).unwrap() {
```

Comment:
nit: binding this to the conditional above so that we can unwrap is OK but seems sloppy. Is there a way we can design this api to not require implicit checks? 

Maybe Could we instead get the visibility and match? Or is there an alternative here?